src/main/groovy/org/pih/warehouse/api/StockMovement.groovy

```diff
@@ -471,34 +475,17 @@ class StockMovement implements Validateable{
         ]
     }
 
-    static def getApplicationTagLib() {
-        return Holders.grailsApplication.mainContext.getBean('org.grails.plugins.web.taglib.ApplicationTagLib')
-    }
-
-    // Because this function needs to handle more than one enum, I left the argument's type as def
-    String getTranslatedDisplayStatus(def status) {
-        return applicationTagLib.message(code: 'enum.' + status?.class?.simpleName + '.' + status)
-    }
-
-    // Mapping statuses for display for the requestor's dashboard
-    // We want to display all statuses from approval workflow (approved, rejected and pending approval)
-    // We want to map all statuses from depot side to "in progress".
-    String getDisplayStatus() {
-        switch(requisition?.status) {
-            case RequisitionStatus.APPROVED:
-                return getTranslatedDisplayStatus(StockMovementStatusCode.APPROVED)
-            case RequisitionStatus.REJECTED:
-                return getTranslatedDisplayStatus(StockMovementStatusCode.REJECTED)
-            case RequisitionStatus.PENDING_APPROVAL:
-                return getTranslatedDisplayStatus(StockMovementStatusCode.PENDING_APPROVAL)
-            case RequisitionStatus.VERIFYING:
-            case RequisitionStatus.PICKING:
-            case RequisitionStatus.PICKED:
-            case RequisitionStatus.CHECKING:
-                return getTranslatedDisplayStatus(StockMovementStatusCode.IN_PROGRESS)
-            default:
-                return getTranslatedDisplayStatus(shipment?.status?.code)
-        }
+    @Deprecated
+    Map<String, String> getDisplayStatus() {
```

Comment:
this is really stretching the limit of what a transient field should be (which is a very simple computation such as taking a timestamp field and converting it to a date). Every time we need to compute this value we're  doing 5 additional lookups, which is really heavy.

That said, I don't know what GORM is doing behind the scenes. If it's pre-fetching all of these foreign keys when we fetch the stock movement itself and then re-using them here, then it's probably fine. I'm just a teeny bit scared that it's not that smart and is doing an extra lookup for each of these on top of the regular lookup when fetching them for the stock movement fields. I don't know if there's any easy way to verify that though

Reply:
> This is really stretching the limit of what a transient field should be (which is a very simple computation such as taking a timestamp field and converting it to a date).  

Honestly, a transient property is just a way to indicate that "the return value of this method is derived and we don't need to store this property in the database". 

Since StockMovement is a command class, we might not even need to use transient. The only reason to use it within a command class would be to avoid having to write a setter method, I guess. 

Transient properties can have getter methods that are as lightweight or heavyweight as we need them to be but with the knowledge that a method that does a lot of database access should probably not be referenced in a haphazard manner that causes N+1 or M(N+1) problems. For example, invoking the method in the StockMovement.toJson() method and then calling an API that fetches and renders a list of all stock movements in the system. 

> Every time we need to compute this value we're doing 5 additional lookups, which is really heavy.

In this case, I don't see 5 additional lookups. I assume you mean the objects that are passed to the context. These objects have already been loaded into memory and there's not much of a chance of any access to lazy properties within those objects. So I think we should be good. But it would definitely be smart to make sure we're not doing anything dumb here.

Reply:
> That said, I don't know what GORM is doing behind the scenes. If it's pre-fetching all of these foreign keys when we fetch the stock movement itself and then re-using them here, then it's probably fine.

Yes, that's exactly what we're doing. Stock Movement is not a domain. It's a command class with references that have been fully hydrated by the time we're doing anything related to status rendering. But you're correct. If there is a lazy access on a property somewhere in the code we're using to render the status, then we should pre-fetch that data.

Reply:
> I don't know if there's any easy way to verify that though

At the moment, you could use a debugger or enable SQL logging (either in the app or the MySQL general log) to see what SQL is being generated. 

It would be amazing if there was a developer plugin like the Django debug toolbar that provided this information in the page as we're developing. 

There's also possibly a way to allow our NewRelic integration to provide some of this information while we are in dev mode. Right now, NewRelic is mainly used in production (and maybe some staging / dev servers). Another option might be to look to see if Sentry's APM provides a more developer-focused experience.

Reply:
@EWaterman we are safe, but it's good that you care and point out such things. Basicallly @jmiranda is right - those should be lazily loaded while getting the stockMovement (e.g. in the `def show` method). Invoking some of the properties (like using them as properties of the helper class, which you pointed out) don't change anything (none SQL should be run).
It will make N+1 (do additional call to any of the properties - `order`/`shipment` etc.) whenever you try to access them "deeper", e.g. in this switch/case:
```groovy
private static RequisitionStatus getOutboundReturnDisplayStatus(Order order) {
           switch (order?.status) {
              case OrderStatus.PENDING:
              case OrderStatus.PLACED:
                  return RequisitionStatus.CREATED
              case OrderStatus.APPROVED:
                  return RequisitionStatus.PICKING
              case OrderStatus.CANCELED:
                  return RequisitionStatus.CANCELED
              case OrderStatus.PARTIALLY_RECEIVED:
              case OrderStatus.RECEIVED:
              case OrderStatus.COMPLETED:
                  return RequisitionStatus.ISSUED
              case OrderStatus.REJECTED:
                  return RequisitionStatus.CANCELED
              default:
                  return RequisitionStatus.PENDING
          }
  
    }
```
here, we would make the fetch for the `order`, as we access a property of an order.

It should not be expensive anyway, since we should touch only 10 SMs at one time (default page size of a list page) and reducing the N+1 would not make sense at the moment, since we would probably have to join fetch all the associations, as we don't know, which of them would be used for a particular context (e.g. for outbound return, we would use `order`, whereas for an inbound return we would use the `shipment`, so there is no sense to join fetch `shipment` for outbound return and vice versa, and it's impossible to determine what would be needed at the level of initialization of the helper instance.)

If you feel like you still didn't get it in 100%, let me know so I could record a video for you with sql logging to show, what I mean here.

Reply:
thanks for the in depth explanation guys!